The job **`x86_64-gnu-llvm-19-3`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12976897079/job/36189696738) [(plain)](https://github.com/rust-lang-ci/rust/commit/4a926fa76050a5f1b75fc5f90a25b9023dd3d8a9/checks/36189696738/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] tests/ui/async-await/async-closures/post-mono-higher-ranked-hang-2.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/checkout/tests/ui/async-await/async-closures/post-mono-higher-ranked-hang-2.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/async-closures/post-mono-higher-ranked-hang-2" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2021"
--- stderr -------------------------------
--- stderr -------------------------------
error: reached the recursion limit while instantiating `recur::<{async closure@/checkout/tests/ui/async-await/async-closures/post-mono-higher-ranked-hang-2.rs:13:24: 13:32}>`
##[error]  --> /checkout/tests/ui/async-await/async-closures/post-mono-higher-ranked-hang-2.rs:13:17
   |
LL |           let _ = recur(&async || {
LL | |             let _ = closure();
LL | |         });
   | |__________^
   |
   |
note: `recur` defined here
  --> /checkout/tests/ui/async-await/async-closures/post-mono-higher-ranked-hang-2.rs:10:1
   |
LL | fn recur<'l>(closure: &'l impl AsyncFn()) -> Pin<Box<dyn Future<Output = ()> + 'l>> {

error: aborting due to 1 previous error
------------------------------------------

```

</details>

--- a/compiler/rustc_type_ir/src/ty_kind/closure.rs
+++ b/compiler/rustc_type_ir/src/ty_kind/closure.rs
@@ -3,9 +3,10 @@ use std::ops::ControlFlow;
 use derive_where::derive_where;
 use rustc_type_ir_macros::{Lift_Generic, TypeFoldable_Generic, TypeVisitable_Generic};
 
+use crate::data_structures::DelayedMap;
 use crate::fold::{TypeFoldable, TypeFolder, TypeSuperFoldable, shift_region};
 use crate::inherent::*;
-use crate::visit::{TypeSuperVisitable, TypeVisitable, TypeVisitor};
+use crate::visit::{TypeSuperVisitable, TypeVisitable, TypeVisitableExt, TypeVisitor};
 use crate::{self as ty, Interner};
 
 /// A closure can be modeled as a struct that looks like:
@@ -471,6 +472,7 @@ impl<I: Interner> CoroutineClosureSignature<I> {
                         interner: cx,
                         region: env_region,
                         debruijn: ty::INNERMOST,
+                        cache: Default::default(),
                     });
                 Ty::new_tup_from_iter(
                     cx,
@@ -498,11 +500,27 @@ struct FoldEscapingRegions<I: Interner> {
     interner: I,
     debruijn: ty::DebruijnIndex,
     region: I::Region,
+
+    // Depends on `debruijn` because we may have types with regions of different
+    // debruijn depths depending on the binders we've entered.
+    cache: DelayedMap<(ty::DebruijnIndex, I::Ty), I::Ty>,
 }
 
 impl<I: Interner> TypeFolder<I> for FoldEscapingRegions<I> {
     fn cx(&self) -> I {
         self.interner
+    }
+
+    fn fold_ty(&mut self, t: I::Ty) -> I::Ty {
+        if !t.has_vars_bound_at_or_above(self.debruijn) {
+            t
+        } else if let Some(&t) = self.cache.get(&(self.debruijn, t)) {
+            t
+        } else {
+            let res = t.super_fold_with(self);
+            assert!(self.cache.insert((self.debruijn, t), res));
+            res
+        }
     }
 
     fn fold_binder<T>(&mut self, t: ty::Binder<I, T>) -> ty::Binder<I, T>

--- a/tests/ui/async-await/async-closures/post-mono-higher-ranked-hang-2.rs
+++ b/tests/ui/async-await/async-closures/post-mono-higher-ranked-hang-2.rs
@@ -1,0 +1,23 @@
+//@ edition: 2021
+//@ build-fail
+
+// Regression test for <https://github.com/rust-lang/rust/issues/135780>.
+
+use std::future::Future;
+use std::ops::AsyncFn;
+use std::pin::Pin;
+
+fn recur<'l>(closure: &'l impl AsyncFn()) -> Pin<Box<dyn Future<Output = ()> + 'l>> {
+    Box::pin(async move {
+        let _ = closure();
+        let _ = recur(&async || {
+            //~^ ERROR reached the recursion limit
+            let _ = closure();
+        });
+    })
+}
+
+fn main() {
+    let closure = async || {};
+    let _ = recur(&closure);
+}

--- a/tests/ui/async-await/async-closures/post-mono-higher-ranked-hang-2.stderr
+++ b/tests/ui/async-await/async-closures/post-mono-higher-ranked-hang-2.stderr
@@ -1,0 +1,18 @@
+error: reached the recursion limit while instantiating `recur::<{async closure@$DIR/post-mono-higher-ranked-hang-2.rs:13:24: 13:32}>`
+  --> $DIR/post-mono-higher-ranked-hang-2.rs:13:17
+   |
+LL |           let _ = recur(&async || {
+   |  _________________^
+LL | |
+LL | |             let _ = closure();
+LL | |         });
+   | |__________^
+   |
+note: `recur` defined here
+  --> $DIR/post-mono-higher-ranked-hang-2.rs:10:1
+   |
+LL | fn recur<'l>(closure: &'l impl AsyncFn()) -> Pin<Box<dyn Future<Output = ()> + 'l>> {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: aborting due to 1 previous error
+

--- a/tests/ui/async-await/async-closures/post-mono-higher-ranked-hang.rs
+++ b/tests/ui/async-await/async-closures/post-mono-higher-ranked-hang.rs
@@ -1,0 +1,63 @@
+//@ build-fail
+//@ aux-build:block-on.rs
+//@ edition:2021
+
+// Regression test for <https://github.com/rust-lang/rust/issues/135780>.
+
+extern crate block_on;
+
+use std::future::Future;
+use std::ops::AsyncFnMut;
+use std::pin::{Pin, pin};
+use std::task::*;
+
+trait Db {}
+
+impl Db for () {}
+
+struct Env<'db> {
+    db: &'db (),
+}
+
+#[derive(Debug)]
+enum SymPerm<'db> {
+    Dummy(&'db ()),
+    Apply(Box<SymPerm<'db>>, Box<SymPerm<'db>>),
+}
+
+pub struct ToChain<'env, 'db> {
+    db: &'db dyn crate::Db,
+    env: &'env Env<'db>,
+}
+
+impl<'env, 'db> ToChain<'env, 'db> {
+    fn perm_pairs<'l>(
+        &'l self,
+        perm: &'l SymPerm<'db>,
+        yield_chain: &'l mut impl AsyncFnMut(&SymPerm<'db>),
+    ) -> Pin<Box<dyn std::future::Future<Output = ()> + 'l>> {
+        Box::pin(async move {
+            match perm {
+                SymPerm::Dummy(_) => yield_chain(perm).await,
+                SymPerm::Apply(l, r) => {
+                    self.perm_pairs(l, &mut async move |left_pair| {
+                        //~^ ERROR reached the recursion limit while instantiating
+                        self.perm_pairs(r, yield_chain).await
+                    })
+                    .await
+                }
+            }
+        })
+    }
+}
+
+fn main() {
+    block_on::block_on(async {
+        let pair = SymPerm::Apply(Box::new(SymPerm::Dummy(&())), Box::new(SymPerm::Dummy(&())));
+        ToChain { db: &(), env: &Env { db: &() } }
+            .perm_pairs(&pair, &mut async |p| {
+                eprintln!("{p:?}");
+            })
+            .await;
+    });
+}

--- a/tests/ui/async-await/async-closures/post-mono-higher-ranked-hang.stderr
+++ b/tests/ui/async-await/async-closures/post-mono-higher-ranked-hang.stderr
@@ -1,0 +1,21 @@
+error: reached the recursion limit while instantiating `ToChain::<'_, '_>::perm_pairs::<{async closure@$DIR/post-mono-higher-ranked-hang.rs:43:45: 43:67}>`
+  --> $DIR/post-mono-higher-ranked-hang.rs:43:21
+   |
+LL | /                     self.perm_pairs(l, &mut async move |left_pair| {
+LL | |
+LL | |                         self.perm_pairs(r, yield_chain).await
+LL | |                     })
+   | |______________________^
+   |
+note: `ToChain::<'env, 'db>::perm_pairs` defined here
+  --> $DIR/post-mono-higher-ranked-hang.rs:34:5
+   |
+LL | /     fn perm_pairs<'l>(
+LL | |         &'l self,
+LL | |         perm: &'l SymPerm<'db>,
+LL | |         yield_chain: &'l mut impl AsyncFnMut(&SymPerm<'db>),
+LL | |     ) -> Pin<Box<dyn std::future::Future<Output = ()> + 'l>> {
+   | |____________________________________________________________^
+
+error: aborting due to 1 previous error
+

